Add postcss-apply plugin

pascalduez opened the pull request:
Hi,

I'm not 100% sure about the plugin order, since it's supposed to be really similar with the custom properties one, I placed it next. 
So  basically all other transformations will happen on the applied rules, instead of the stored rules. Maybe in case of huge sets usage it could have an impact? 

I added [tests locally](https://github.com/pascalduez/postcss-apply/blob/master/test/integration.test.js) to the plugin to ensure it doesn't interfere with the custom properties one.

Refs #203

kevinSuttle commented:
Looking forward to this landing.

ben-eb commented on docs/content/index.md:
```diff
@@ -54,6 +54,9 @@ title: cssnext - Use tomorrow’s CSS syntax, today.
         <a href="/features/#custom-properties-var">custom properties & <code>var()</code></a>
       </li>
       <li class="r-Grid-cell r-minS--1of2">
+        <a href="/features/#custom-properties-set-apply">custom properties set & <code>@apply</code></a>
```
`&amp;`?

pascalduez replied:
Replicated from above, but probably better that's right.

ben-eb commented on docs/content/features.md:
````diff
@@ -42,6 +42,30 @@ might happen).
+}
+```
+
+(The same DOMless restrictions than the custom properties plugin applies).
````
than => as

zeke replied:
that?

ben-eb replied:
Yeah, that is probably better. Or it can be rewritten;

> The same DOM restrictions as the custom properties plugin apply.

pascalduez replied:
Thanks! I knew that sentence felt a bit tortured :)

MoOx commented:
Will this works with custom props referenced in the spec like this?

```css
:root {
  --toolbar-theme: {
    background-color: hsl(120, 70%, 95%);
    border-radius: 4px;
    border: 1px solid var(--theme-color late); /* IS THIS SUPPORTED */ ?
  };
  --toolbar-title-theme: {
    color: green;
  };
}

.toolbar {
  @apply --toolbar-theme;
}
.toolbar > .title {
  @apply --toolbar-title-theme;
}
```
I guess it don't since apply transformation is applied after custom props. But maybe I am missing something.

pascalduez commented:
> Will this works with custom props referenced in the spec like this?

I added some tests to the plugin for this very use case, seems okay to me: https://github.com/pascalduez/postcss-apply/blob/master/test/integration.test.js#L136-L227

Custom properties (single) are correctly parsed and replaced inside the custom properties (set), provided the plugins are in this PR order.

MoOx commented:
Looks good enough. Thanks for this feature.

MoOx commented:
Released as 2.7.0.
